pkg/middlewares/aws_api_gateway_rest_api_expander.go

```diff
@@ -0,0 +1,105 @@
+package middlewares
+
+import (
+	"encoding/json"
+
+	"github.com/cloudskiff/driftctl/pkg/resource"
+	"github.com/cloudskiff/driftctl/pkg/resource/aws"
+	"github.com/getkin/kin-openapi/openapi2"
+	"github.com/getkin/kin-openapi/openapi3"
+)
+
+// Explodes api gateway rest api body attribute to dedicated resources as per Terraform documentation (https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/api_gateway_rest_api)
+type AwsApiGatewayRestApiExpander struct {
+	resourceFactory resource.ResourceFactory
+}
+
+func NewAwsApiGatewayRestApiExpander(resourceFactory resource.ResourceFactory) AwsApiGatewayRestApiExpander {
+	return AwsApiGatewayRestApiExpander{
+		resourceFactory: resourceFactory,
+	}
+}
+
+func (m AwsApiGatewayRestApiExpander) Execute(remoteResources, resourcesFromState *[]*resource.Resource) error {
+	newStateResources := make([]*resource.Resource, 0)
+	for _, res := range *resourcesFromState {
+		// Ignore all resources other than aws_api_gateway_rest_api
+		if res.ResourceType() != aws.AwsApiGatewayRestApiResourceType {
+			newStateResources = append(newStateResources, res)
+			continue
+		}
+
+		newStateResources = append(newStateResources, res)
+
+		err := m.handleBody(res, &newStateResources, remoteResources)
+		if err != nil {
+			return err
+		}
+	}
+	*resourcesFromState = newStateResources
+	return nil
+}
+
+func (m *AwsApiGatewayRestApiExpander) handleBody(api *resource.Resource, results, remoteResources *[]*resource.Resource) error {
+	body := api.Attrs.GetString("body")
+	if body == nil || *body == "" {
+		return nil
+	}
+
+	docV3 := &openapi3.T{}
+	if err := json.Unmarshal([]byte(*body), &docV3); err != nil {
+		return err
+	}
+	// It's an OpenAPI v3 document
+	if docV3.OpenAPI != "" {
+		return m.handleBodyV3(docV3, results, remoteResources)
+	}
+
+	docV2 := &openapi2.T{}
+	if err := json.Unmarshal([]byte(*body), &docV2); err != nil {
+		return err
+	}
+	// It's an OpenAPI v2 document
+	if docV2.Swagger != "" {
+		return m.handleBodyV2(docV2, results, remoteResources)
+	}
+
+	return nil
+}
+
+func (m *AwsApiGatewayRestApiExpander) handleBodyV3(doc *openapi3.T, results, remoteResources *[]*resource.Resource) error {
+	for path := range doc.Paths {
+		m.createApiGatewayResource(path, results, remoteResources)
+	}
+	return nil
+}
+
+func (m *AwsApiGatewayRestApiExpander) handleBodyV2(doc *openapi2.T, results, remoteResources *[]*resource.Resource) error {
+	for path := range doc.Paths {
+		m.createApiGatewayResource(path, results, remoteResources)
+	}
+	return nil
+}
```

Comment:
In our initial discussion I was not aware that we do not need details of openapi documents. I think we can get rid of the  need to use an external dependency.

I mean I don't think theses methods are very usefull. We do not care to know if we are on a v2 or v3 openapi document, we only need the keys of the `Paths` map.

What about something like that :

```go
type OpenAPIDocument struct {
	Paths map[string]interface{} `json:"paths"`
}


	input := `
	{
    "info": {
        "title": "example",
        "version": "1.0"
    },
    "openapi": "3.0.1",
    "paths": {
        "/path1": {
            "get": {
                "x-amazon-apigateway-integration": {
                    "httpMethod": "GET",
                    "payloadFormatVersion": "1.0",
                    "type": "HTTP_PROXY",
                    "uri": "https://ip-ranges.amazonaws.com/ip-ranges.json"
                }
            }
        },
        "/path1/path2": {
            "get": {
                "x-amazon-apigateway-integration": {
                    "httpMethod": "GET",
                    "payloadFormatVersion": "1.0",
                    "type": "HTTP_PROXY",
                    "uri": "https://ip-ranges.amazonaws.com/ip-ranges.json"
                }
            }
        }
    }
}
`

	doc := &OpenAPIDocument{}
	if err := json.Unmarshal([]byte(input), doc); err != nil {
		return err
	}

	for path := range doc.Paths {
		fmt.Println(path)
	}


```

Reply:
We do need the details of the rest_api `body` attribute for several "future" resources as per their docs:

![image](https://user-images.githubusercontent.com/8110579/135819029-946eec43-5902-471d-9b4b-392bdd1e2ab6.png)

This middleware will grow to create other resources than just `aws_api_gateway_resource `

Reply:
Ok I see 👍🏻